Sources/CoreMetrics/Metrics.swift

```diff
@@ -488,6 +580,74 @@ public protocol MetricsFactory {
     func destroyTimer(_ handler: TimerHandler)
 }
 
+/// Wraps a CounterHandler, adding support for incrementing by floating point values by storing an accumulated floating point value and recoding increments to the underlying CounterHandler after crossing integer boundaries.
+private class AccumulatingRoundingFloatingPointCounter: FloatingPointCounterHandler {
+    private static let kAutoStateResetThreshold = Double(Int64.max - (2 << 8))
+
+    private var lock = Lock()
+    private var rawValue: Double = 0
+    private var roundedValue: Int64 = 0
+    private(set) var counterHandler: CounterHandler
+
+    init(counterHandler: CounterHandler) {
+        self.counterHandler = counterHandler
+    }
+
+    func increment(by amount: Double) {
+        assert(amount < AccumulatingRoundingFloatingPointCounter.kAutoStateResetThreshold)
+        self.lock.withLockVoid {
+            let incremented = rawValue + amount
+            if incremented < AccumulatingRoundingFloatingPointCounter.kAutoStateResetThreshold {
+                self.rawValue = incremented
+            } else {
+                self.rawValue = amount
+                self.roundedValue = 0
+            }
+            let roundedIncrement = Int64(rawValue) - roundedValue
+            assert(roundedIncrement >= 0)
+            if roundedIncrement > 0 {
+                self.roundedValue += roundedIncrement
+                self.counterHandler.increment(by: roundedIncrement)
+            }
+        }
+    }
+
+    func reset() {
+        self.lock.withLockVoid {
+            self.rawValue = 0
+            self.roundedValue = 0
+            self.counterHandler.reset()
```

Comment:
self.counterHandler.reset() outside of the lock?

Reply:
That could interleave with calls to `increment` though unlikely, right?

Reply:
my is worry about a deadlock if the handler gets stuck for whatever reason. presumably, the handler can handle its own concurrency so we dont need to protect it?

Reply:
I don't think that is thread safe:

``` swift
 Time | Thread A                 | Thread B
------+--------------------------+-------------------
    0 | counter.inc(...)         | counter.reset()
    1 | <blocked>                | lock
    2 | <blocked>                | fraction = 0
    3 | <blocked>                | unlock
    4 | lock                     | <not scheduled>
    5 | fraction = x             | <not scheduled>
    6 | handler.inc(...)         | <not scheduled>
    7 | unlock                   | <not scheduled>
    8 | return                   | handler.reset()
    9 |                          | return
```

Depending on the value passed to `inc` and the value in `fraction` you could get incorrect results I think.

Reply:
oh, I think you are worried about correctness, not actual concurrent access, right?

Reply:
Yep, it seems like a correctness issue to not lock around the reset